KAFKA-5849: Add process stop, round trip workload, partitioned test

cmccabe opened the pull request:
* Implement process stop faults via SIGSTOP / SIGCONT

* Implement RoundTripWorkload, which both sends messages, and confirms that they are received at least once.

* Allow Trogdor tasks to block until other Trogdor tasks are complete.

* Add CreateTopicsWorker, which can be a building block for a lot of tests.

* Simplify how TaskSpec subclasses in ducktape serialize themselves to JSON.

* Implement some fault injection tests in round_trip_workload_test.py

ijuma commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RoundTripWorker.java:
```diff
@@ -0,0 +1,328 @@
+            final ToSendTracker toSendTracker = new ToSendTracker(spec.maxMessages());
+            long messagesSent = 0;
+            long uniqueMessagesSent = 0;
+            log.debug("{}: Starting RoundTripWorker#ProducerRunnable.", id);
```
Is there a reason why you're not using `LogContext` instead of manually adding the `id` in every log message?

cmccabe replied:
Trogdor hasn't been converted to using LogContexts yet.  It does seem like it would be a good idea, though.  Is it worth doing in a separate PR?

ijuma replied:
Sounds good.

rajinisivaram commented on tests/kafkatest/services/trogdor/process_stop_fault_spec.py:
```diff
@@ -0,0 +1,38 @@
+from kafkatest.services.trogdor.task_spec import TaskSpec
+
+
+class ProcessStopFaultSpec(TaskSpec):
```
Would it be better to call this `JavaProcessStopFaultSpec` to make it obvious that it is meant. only for Java processes?

cmccabe replied:
The intention is to add support for stopping non-java processes later.  We can add a field like 'processName' to go along with 'javaProcessName'.

rajinisivaram commented on tests/kafkatest/services/trogdor/task_spec.py:
```diff
@@ -25,21 +25,32 @@ class TaskSpec(object):
-        raise NotImplemented
+        node_names = []
+        for obj in nodes:
+            if isinstance(obj, basestring):
```
Why? Do we set nodenames in `nodes`?

cmccabe replied:
It's a convenience to allow people to set up TaskSpecs using either Node objects, or the names of nodes.

rajinisivaram commented on tests/kafkatest/tests/core/round_trip_workload_test.py:
```diff
@@ -0,0 +1,99 @@
+from kafkatest.services.zookeeper import ZookeeperService
+
+
+class RoundTripWorkloadTest(Test):
```
When I saw the test, I assumed it was a good path workload test. But actually it is injecting faults, so perhaps that could be reflected in the class name?

cmccabe replied:
Good point.  I renamed this to "RoundTripFaultTest"

rajinisivaram commented on tests/kafkatest/tests/core/round_trip_workload_test.py:
```diff
@@ -0,0 +1,99 @@
+                                     self.workload_service.client_node,
+                                     self.workload_service.bootstrap_servers,
+                                     10000,
+                                     100000,
```
Perhaps use named arguments for these two integers?

rajinisivaram commented on tests/kafkatest/tests/core/round_trip_workload_test.py:
```diff
@@ -0,0 +1,99 @@
+                                     deps=["create_topic"])
+
+    def setUp(self):
+        self.trogdor.start()
```
Perhaps start trogdor after starting ZK and Kafka so that it is the reverse order of teardown?

rajinisivaram commented on tests/kafkatest/tests/core/round_trip_workload_test.py:
```diff
@@ -0,0 +1,99 @@
+        workload1.wait_for_done(timeout_sec=720)
+        stop1.stop()
+        stop1.wait_for_done()
+        time.sleep(1)
```
Why? Can you add a comment?

cmccabe replied:
The last sleep shouldn't be there-- will remove.

rajinisivaram commented on tests/kafkatest/tests/core/round_trip_workload_test.py:
```diff
@@ -0,0 +1,99 @@
+    def test_round_trip_workload(self):
+        self.trogdor.create_task("create_topic", self.create_topics_spec)
+        workload1 = self.trogdor.create_task("workload1", self.round_trip_spec)
+        workload1.wait_for_done(timeout_sec=720)
```
Is there a reason for the 12 minute timeout in these tests (I mean why not 5 minutes or 10 minutes)? Doesn't actually matter, wondering if it was related to some config?

cmccabe replied:
It was just an arbitrary choice.  I will make it 10 minutes instead.

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/common/WorkerUtils.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public static void abort(String what, Throwable e, KafkaFutureImpl<String> doneFuture)
+            throws KafkaException {
+        log.warn(what + " caught an exception: ", e);
```
Avoid concatenation? This should still log the exception with stack trace: `log.warn("{} caught an exception: ", what, e)`

rajinisivaram commented on tools/src/main/java/org/apache/kafka/trogdor/workload/CreateTopicsWorker.java:
```diff
@@ -0,0 +1,86 @@
+        try (AdminClient adminClient = AdminClient.create(props)) {
+            List<String> topicsToCreate = new ArrayList<>();
+            for (int i = 0; i < spec.numTopics(); i++) {
+                topicsToCreate.add(WorkerUtils.topicIndexToName(i));
```
Just to clarify - do we always expect a single worker that creates all topics for.a test at the same time?

cmccabe replied:
No, some tests should create topics over time.  But I think this is fine for round trip test.

rajinisivaram replied:
I asked because there is no base name or starting index for the topic. Since this code is not specific to a test, perhaps, it ought to have either a base name or the index to start from?

rajinisivaram reviewed (commented):
@cmccabe Thank you for the PR. Looks good overall, left a few minor comments.

I ran the tests locally and saw a failure in `test_round_trip_workload_with_broker_partition`. Reran the test and it was fine. Perhaps there is a transient timing issue?

```
workload1 failed to finish in the expected amount of time.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/ducktape/tests/runner_client.py", line 132, in run
    data = self.run_test()
  File "/usr/local/lib/python2.7/dist-packages/ducktape/tests/runner_client.py", line 185, in run_test
    return self.test_context.function(self.test)
  File "/opt/kafka-dev/tests/kafkatest/tests/core/round_trip_workload_test.py", line 70, in test_round_trip_workload_with_broker_partition
    workload1.wait_for_done(timeout_sec=720)
  File "/opt/kafka-dev/tests/kafkatest/services/trogdor/trogdor.py", line 332, in wait_for_done
    err_msg="%s failed to finish in the expected amount of time." % self.id)
  File "/usr/local/lib/python2.7/dist-packages/ducktape/utils/util.py", line 36, in wait_until
    raise TimeoutError(err_msg)
TimeoutError: workload1 failed to finish in the expected amount of time.
```

rajinisivaram commented:
There was also a warning during teardown of `test_produce_consume_with_broker_pause`. Is that expected?
```
kafkatest.tests.core.round_trip_workload_test.RoundTripWorkloadTest.test_produce_consume_with_broker_pause: Error running teardown method: Kafka node failed to stop
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/ducktape/tests/runner_client.py", line 189, in _do_safely
    action()
  File "/opt/kafka-dev/tests/kafkatest/tests/core/round_trip_workload_test.py", line 53, in teardown
    self.kafka.stop()
  File "/usr/local/lib/python2.7/dist-packages/ducktape/services/service.py", line 278, in stop
    self.stop_node(node)
  File "/opt/kafka-dev/tests/kafkatest/services/kafka/kafka.py", line 268, in stop_node
    wait_until(lambda: len(self.pids(node)) == 0, timeout_sec=60, err_msg="Kafka node failed to stop")
  File "/usr/local/lib/python2.7/dist-packages/ducktape/utils/util.py", line 36, in wait_until
    raise TimeoutError(err_msg)
TimeoutError: Kafka node failed to stop
```

cmccabe commented:
> There was also a warning during teardown of test_produce_consume_with_broker_pause. Is that expected?

I looked into this and found that the paused-and-then-resumed broker was successfully doing a controlled shutdown, but that its NetworkClient was waiting for a TCP session to time out before shutting down.  I will add a 'kill -9' to the test to avoid having this warning spewed to the output.

cmccabe commented:
> I ran the tests locally and saw a failure in test_round_trip_workload_with_broker_partition. Reran the test and it was fine. Perhaps there is a transient timing issue?

Do you have the logs from the run when it failed?

cmccabe committed:
KAFKA-5849: Add process stop, round trip workload, partitioned test

* Implement process stop faults via SIGSTOP / SIGCONT

* Implement RoundTripWorkload, which both sends messages, and confirms
that they are received at least once.

* Simplify how TaskSpec subclasses in ducktape serialize themselves to
JSON.

* Implement some fault injection tests in round_trip_workload_test.py

cmccabe commented:
There was an internal jenkins error:
```
/tmp/jenkins6224879222694533419.sh: line 4: /home/jenkins/tools/gradle/3.5/bin/gradle: No such file or directory
```

cmccabe commented:
>> I ran the tests locally and saw a failure in test_round_trip_workload_with_broker_partition. Reran the test and it was fine. Perhaps there is a transient timing issue?
>
> Do you have the logs from the run when it failed?

I looked at the logs, and it appears you hit a race condition in the test.  I fixed it

cmccabe commented:
I reworked this change a little bit to avoid having a separate task to create the tasks.  It doesn't seem necessary here, and it's simpler without it.  I also fixed a race condition in the test that could lead to the test appearing to fail even though all messages were sent and received correctly.

cmccabe commented:
Got this again:
```
/tmp/jenkins4792976252943520186.sh: line 4: /home/jenkins/tools/gradle/3.5/bin/gradle: No such file or directory
```

cmccabe commented:
retest this please

rajinisivaram commented:
retest this please

rajinisivaram commented:
@cmccabe There is an unused import in `RoundTripWorker.java` causing checkstyle failure. Can you fix that, please?

cmccabe commented:
fixed checkstyle
